### PR TITLE
Fix "File/Harness status"

### LINKED ISSUE #419: "File status" still shown for testharness.js tests
A follow-up to https://github.com/web-platform-tests/wpt.fyi/issues/371.

"File status" is still shown for these test:
https://staging.wpt.fyi/results/compat/webkit-text-fill-color-currentColor.html
https://staging.wpt.fyi/results/css/css-flexbox/flexbox_first-letter.html
https://staging.wpt.fyi/results/dom/historical.html

Probably all tests are affected.

### PR DESCRIPTION
**Pull request by Hexcles**
In #405, we turned `test-file-results` into a thin layer was only responsible for fetching data and delegated the UI logic to new embedded components `test-file-results-table-{terse,verbose}` (which were subclasses of `AbstractTestFileResultsTable(TestRunsBase)`).

Therefore, information needs to be explicitly passed into the new components. Yet we forgot to pass `isTestHarness`, which caused the regression. This PR turns `isTestHarness` into `hasSubtests` (which is really a more accurate name as we don't know the type of a test for sure without consulting the manifest), and passes the property down to the embedded components.

This PR also removes the `is` getter from `abstract-test-file-results-table`; the abstract class is not a DOM template and doesn't need the `is` getter.

Fixes #419 .

## Review Information

Check out a testharness test and a reftest in the staging instance. They should show "Harness status" and "File status" respectively.

### CODE REVIEW

File: webapp/components/abstract-test-file-results-table.html [hunk 1]
```diff
@@ -8,19 +8,73 @@
...
+          },
         };
       }
 
+      static get observers() {
+        return ['fetchTestFile(path, testRuns)'];
+      }
+
+      async fetchTestFile(path, testRuns) {
+        if (!path || !testRuns) {
```
Comment (foolip) [inline webapp/components/abstract-test-file-results-table.html]: `this.subtestNames = []; // Clear any existing rows` before this wasn't copied over. Is it no longer possible that `this.subtestNames` should become empty after being non-empty? I haven't read too carefully to discover the invariants, sorry...

File: webapp/components/abstract-test-file-results-table.html [hunk 2]
```diff
@@ -8,19 +8,73 @@
...
+        const response = await window.fetch(url);
+        if (!response.ok) {
+          return null;
+        }
+        return response.json();
+      }
+
+      resultsURL(testRun, path) {
+        path = this.encodeTestPath(path);
+        // This is relying on the assumption that result files end with '-summary.json.gz'.
```
Comment (foolip) [inline webapp/components/abstract-test-file-results-table.html]: Aside: I don't think anyone will be too sad if @mdittmer makes these lines go *poof* :)

File: webapp/components/abstract-test-file-results-table.html [hunk 3]
```diff
@@ -8,19 +8,73 @@
...
+        resultPerTestRun.forEach((resultData, i) => {
+          if (!resultData) {
+            testRuns[i].subtests = {};
+            testRuns[i].subtests['STATUS'] = { status: '(results not found)' };
+            return;
+          }
+          testRuns[i].subtests = {};
+          testRuns[i].subtests['STATUS'] = { status: resultData.status };
+          if (['OK', 'ERROR'].includes(resultData.status)) {
+            isTestHarness = true;
```
Comment (foolip) [inline webapp/components/abstract-test-file-results-table.html]: I can't quite tell if the fix is around here and hard to spot because of the moved code, or if the refactoring itself fixes this as a side effect. Either works, but if there are hard-to-spot changes here, two commits would be appreciated for a nervous reviewer :)

**Comment by foolip** [comment]
Fix confirmed with https://fix-status-dot-wptdashboard-staging.appspot.com/results/compat/webkit-text-fill-color-currentColor.html

File: webapp/components/abstract-test-file-results-table.html [hunk 4]
```diff
@@ -8,19 +8,74 @@
   <script>
```
Comment (mdittmer) [inline webapp/components/abstract-test-file-results-table.html]: Broad point of clarification: Aren't these tables usually embedded in the context of other components that perform fetch? Are we sure that this addition isn't double-fetching for some view configurations?

File: webapp/components/abstract-test-file-results-table.html (hunk 1 above)
Comment (Hexcles) [reply webapp/components/abstract-test-file-results-table.html]: I think this is no longer needed, but in case I missed something, I added it back, but to a different place: inside the `if` block right before `return`. If the condition doesn't hold, `this.subtestNames` will always be updated at the end of this function.

File: webapp/components/abstract-test-file-results-table.html (hunk 3 above)
Comment (Hexcles) [reply webapp/components/abstract-test-file-results-table.html]: No, the real fix is the two new lines in `test-file-results.html`. I've split the changes into two commits.

**Comment by Hexcles** [comment]
@mdittmer just pointed out that my first commit (the refactor) causes the frontend to fetch individual test results a second time when the user clicks the toggle...

File: webapp/components/abstract-test-file-results-table.html (hunk 4 above)
Comment (Hexcles) [reply webapp/components/abstract-test-file-results-table.html]: Thanks for spotting this! Although my refactor might be a little bit cleaner logically, we don't want double fetching. I just reverted my refactor. Let's fix the bug first.

**Comment by Hexcles** [comment]
Ping @mdittmer ; this is now much shorter (just a fix without refactoring). Does it look good to you?

**Comment by Hexcles** [comment]
Sorry please disregard... This apparently doesn't fix this issue. My original patch worked in a way I didn't fully understand...

**Comment by Hexcles** [comment]
Sorry for the back and forth. I misunderstood the code I was trying to fix. This should now be a cleaner fix that actually works. I've also updated the PR description. PTAL, @foolip & @mdittmer . Thanks!

File: webapp/components/abstract-test-file-results-table.html [hunk 5]
```diff
@@ -8,19 +8,30 @@
...
 
+      computeIsTestHarness(subtestNames) {
+        // subtestNames is always updated after testRuns[].subtests are updated,
+        // so as long as 'STATUS' exists in subtestNames, we can safely access
+        // testRuns[].subtests['STATUS'].
+        if (subtestNames.includes('STATUS')) {
+          return false;
+        }
+        const statuses = this.testRuns.map(tr => tr.subtests['STATUS'].status);
+        return statuses.includes('OK') || statuses.includes('ERROR');
```
Comment (foolip) [inline webapp/components/abstract-test-file-results-table.html]: This all seems roundabout to me. Why not just check if there are any subtests? And if there aren't, it might be a single harness error, but I think we'd need the manifest to know for sure what kind of test it was? "TIMEOUT" might be either kind, for example.

Comment (Hexcles) [reply webapp/components/abstract-test-file-results-table.html]: There's always at least a subtest as we currently spoof `testRuns[].subtests` to always include a fake `STATUS` subtest for the file. We could check if there's other non-'STATUS' subtest, but that would fail if the test is a testharness test and all browsers produce harness errors. Eventually we should rely on manifest. The search API may give us the ability to do more useful things on the server side.

Comment (mdittmer) [reply webapp/components/abstract-test-file-results-table.html]: Where do these data come from? From our sharded individual test files? If so, I believe they contain separate `"status": ...` and `"subtest": [ ... status ... ]`. Perhaps the better long term solution (this may have been in your previous refactor, I'm not sure) would be to disentangle the test status and subtest statuses in whatever component processes data from GCS.

Comment (mdittmer) [reply webapp/components/abstract-test-file-results-table.html]: ... to be clear, I would be fine with filing an issue to handle that and landing this code.

**Comment by Hexcles** [comment]
The file-level status is prepended to subtests as a special subtest called "STATUS" here:
https://github.com/web-platform-tests/wpt.fyi/blob/33fcde2fcbf0d21ad698c5b774060fd086952ef0/webapp/components/test-file-results.html#L110

And I agree this isn't pretty.

**Comment by foolip** [comment]
Long comment so hoisting out of review thread.

When we load https://staging.wpt.fyi/results/compat/webkit-text-fill-color-currentColor.html, the data for Chrome is from https://storage.googleapis.com/wptd-staging/2dda7b8c10c7566fa6167a32b09c85d51baf2a85/chrome-68.0.3440.106-linux-16.04-edf200244e/compat/webkit-text-fill-color-currentColor.html. Verbatim response body: `{"test": "/compat/webkit-text-fill-color-currentColor.html", "status": "OK", "message": null, "duration": 118, "subtests": [{"status": "PASS", "message": null, "name": "Test that currentColor is not affected by -webkit-text-fill-color"}]}`

A reftest, https://storage.googleapis.com/wptd-staging/2dda7b8c10c7566fa6167a32b09c85d51baf2a85/chrome-68.0.3440.106-linux-16.04-edf200244e/compat/webkit-text-fill-color-property-001a.html, however, has this response body: `{"test": "/compat/webkit-text-fill-color-property-001a.html", "status": "PASS", "message": null, "duration": 536, "subtests": []}`

And here's a webdriver response body from https://staging.wpt.fyi/results/webdriver/tests/accept_alert/accept.py: `{"test": "/webdriver/tests/accept_alert/accept.py", "status": "OK", "message": null, "duration": 3047, "subtests": [{"status": "PASS", "message": null, "name": "test_null_response_value"}, {"status": "PASS", "message": null, "name": "test_no_browsing_context"}, {"status": "PASS", "message": null, "name": "test_no_user_prompt"}, {"status": "PASS", "message": null, "name": "test_accept_alert"}, {"status": "PASS", "message": null, "name": "test_accept_confirm"}, {"status": "PASS", "message": null, "name": "test_accept_prompt"}]}`

In other words, we can't tell testharness.js from wdspec, but we *can* tell if there are subtests, and when there are we call the the test-level status a "harness status".

@Hexcles, if the data has been mangled by the time it reaches the web component, can file a follow up issue to fix this with a more straightforward test based on the above data formats?

**Comment by foolip** [comment]
> The file-level status is prepended to subtests as a special subtest called "STATUS" here

Yep, that's all part of https://github.com/web-platform-tests/wpt.fyi/issues/62, to stop treating harness status as a test in itself.

For now, perhaps we can simply compute if there are subtests *before* we add the fake one, and use that bool later?

**Comment by Hexcles** [comment]
@foolip good point. See my latest commit.

File: webapp/components/test-file-results.html [hunk 6]
```diff
@@ -108,19 +110,19 @@
...
+              testRuns[i].subtests[subtestResult.name] = subtestResult;
+              if (!subtestNames.includes(subtestResult.name)) {
+                subtestNames.push(subtestResult.name);
+              }
             }
           }
         });
-
-        this.isTestHarness = isTestHarness;
+        this.hasSubtests = true;
```
Comment (foolip) [inline webapp/components/test-file-results.html]: Should be `= hasSubtests`?

File: webapp/components/test-file-results.html [hunk 7]
```diff
@@ -108,19 +110,19 @@
...
-          if (['OK', 'ERROR'].includes(resultData.status)) {
-            isTestHarness = true;
-          }
-
-          for (let subtestResult of resultData.subtests) {
-            testRuns[i].subtests[subtestResult.name] = subtestResult;
-            if (!subtestNames.includes(subtestResult.name)) {
-              subtestNames.push(subtestResult.name);
+          if (resultData.subtests && resultData.subtests.length > 0) {
+            // Since we spoof the file-level status also as a "subtest", we
```
Comment (foolip) [inline webapp/components/test-file-results.html]: test-level. And maybe comment here that this is still imperfect, since a testharness.js test that throws an exception before registering the first test will still say "Test status" instead of "Harness status".

**Comment by foolip** [comment]
Doesn't work as expected at https://fix-status-dot-wptdashboard-staging.appspot.com/results/battery-status/battery-iframe.https.html

What kinds of tests should we have for this? WCT?

**Comment by Hexcles** [comment]
@foolip fixed in the latest commit ("fixup"). Yeah I'll add a web component test for this when I fix the existing tests for this component. The test setup is broken now.

File: webapp/components/abstract-test-file-results-table.html [hunk 8]
```diff
@@ -8,24 +8,23 @@
...
         };
       }
 
-      subtestNameForDisplay(subtestName, isTestHarness) {
+      subtestNameForDisplay(subtestName, hasSubtests) {
         if (subtestName !== 'STATUS') {
           return subtestName;
         }
-        return isTestHarness ? 'Harness status' : 'File status';
+        return hasSubtests ? 'Harness status' : 'File status';
```
Comment (foolip) [inline webapp/components/abstract-test-file-results-table.html]: Wasn't this fixed to "Test status" in this PR already?

Comment (Hexcles) [reply webapp/components/abstract-test-file-results-table.html]: Sorry... I have too many local branches and confused myself when merging. Thanks for double checking on this!
